ext/standard/string.c

```diff
@@ -5010,16 +5010,20 @@ PHP_FUNCTION(parse_str)
 	zval *arrayArg = NULL;
 	char *res = NULL;
 	size_t arglen;
+	bool mergeParams = 0;
 
-	ZEND_PARSE_PARAMETERS_START(2, 2)
+	ZEND_PARSE_PARAMETERS_START(2, 3)
 		Z_PARAM_STRING(arg, arglen)
 		Z_PARAM_ZVAL(arrayArg)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_BOOL(mergeParams)
 	ZEND_PARSE_PARAMETERS_END();
 
 	arrayArg = zend_try_array_init(arrayArg);
 	if (!arrayArg) {
 		RETURN_THROWS();
 	}
+	Z_EXTRA_P(arrayArg) = (uint16_t)mergeParams;
```

Comment:
I would prefer not using `Z_EXTRA_P` for non-VM stuff. It's very easy to run into collisions.

Reply:
I understand the concern, indeed just setting the extra value anywhere else than for PARSE_STRING related code, things break..

I see 3 other options for passing the merge_params flag:
- either add a new constant in php_variables.h like PARSE_STRING_MERGE
- or add a new argument to treat_data function in SAPI.h (void (*treat_data)(int arg, char *str, zval *destArray, bool mergeParams);)
- create a new treat_data function in SAPI

The first option seems hacky/not really appropriate to me, but with less code change.
Second option might need some refactoring where treat_data gets called.
Latest option would bring lots of code duplication :/

If you have other ideas or a preference for an option above let me know :)

Reply:
Considering that it's kind of a different logic, new constant might be more appropriate here.